### PR TITLE
Stage 5C.1: Failure Detection and Recovery State Model

### PR DESCRIPTION
**Pull request by mohavro**
## **User description**

This pull request introduces a comprehensive **Failure Detection and Recovery State Model** for distributed rebuild processes, significantly enhancing the system's robustness and safety.

Key changes include:

1.  **Database Schema Enhancements**:
    *   Added `active_worker_id` and `last_heartbeat_at` columns to the `rebuild_jobs` table to track the worker currently executing a rebuild and its liveness.
    *   Introduced a Python-based migration runner to safely apply schema changes, ensuring idempotency for SQLite.

2.  **Distributed Coordination Logic**:
    *   **Heartbeat Mechanism**: Implemented a heartbeat mechanism in `AssetGraphRepository` to allow workers to signal liveness, enforcing single-worker ownership of running rebuild jobs.
    *   **Lock State Checking**: The `DistributedLock` now includes a `check_state()` method to explicitly verify the lock's status (valid, expired, unknown, lost) against the database.
    *   **Active Rebuild State Retrieval**: The repository can now retrieve the single active rebuild job, raising an error if multiple jobs are found in a `RUNNING` state, indicating a critical inconsistency.

3.  **Failure Detection Module**:
    *   A new `rebuild_failure_detection` module identifies various inconsistent states, such as `STALE_OWNERSHIP` (old heartbeat), `ORPHANED_RUNNING` (DB job running but no active runtime executor), `ZOMBIE_EXECUTOR` (runtime executor active but no DB job), and `CRASH_SUSPICION` (worker assigned but heartbeat missing or stale).

4.  **Recovery State Model & Decision Logic**:
    *   A new `rebuild_recovery` module provides deterministic decision-making for recovery actions based on detected inconsistencies and lock validity.
    *   Defines `RecoveryAction` types: `RESUME` (safe to proceed), `RESET` (state must be reset), `WAIT` (must wait for stabilization), and `UNSAFE` (execution is blocked due to high-risk conditions like split-brain).

5.  **Recovery Gate Enforcement**:
    *   A `RecoveryGate` component is introduced to block rebuild execution if the system is in an unsafe or inconsistent state.
    *   All attempts to initiate a graph rebuild now pass through this gate, which evaluates the current lock, database, and runtime states to determine if execution is safe. An `ExecutionBlockedError` is raised if execution is not safe, preventing operations that could lead to data corruption.

6.  **Monitoring & Observability**:
    *   New Prometheus metrics (`REBUILD_STATE_STATUS` gauge and `REBUILD_RECOVERY_TRIGGERS` counter) provide real-time visibility into the rebuild process's health and detected inconsistencies.
    *   The `REBUILD_STATE_STATUS` metric is initialized from the authoritative database state on application startup, ensuring accurate reporting even after service restarts.

7.  **CI/CD Safety Gates**:
    *   A new GitHub Actions workflow (`ci-gate-spec.yaml`) enforces distributed coordination safety through three stages:
        *   **Static Invariant Analysis**: Custom static analysis ensures critical safety rules are followed (e.g., no direct ownership bypass, `RecoveryGate` always used for rebuilds).
        *   **Multi-Worker Simulation**: Integration tests simulate concurrent worker behavior to validate invariants like single active owner and no overlapping execution.
        *   **Failure Injection Tests**: Dedicated tests simulate various failure modes (lock expiry, worker crashes, DB issues) to validate the system's recovery and resilience.

8.  **API Integration**:
    *   The API's `lifespan` function now initializes the rebuild state metric from the database.
    *   Rebuild endpoints are updated to use the `RecoveryGate` and to update the new rebuild state metric throughout the rebuild lifecycle.
    *   Error handling for `ExecutionBlockedError` is added to API responses.

These changes establish a robust framework for detecting and managing failures in distributed rebuild operations, ensuring data integrity and system reliability under various fault conditions.

___

## **CodeAnt-AI Description**
**Add rebuild recovery checks that block unsafe execution and track worker liveness**

### What Changed
- Rebuilds now record a worker owner and heartbeat so the app can spot stale, crashed, or orphaned running jobs
- Startup now restores rebuild status metrics from the database instead of leaving them blank or outdated after a restart
- Rebuild execution is blocked when the lock, database state, or runtime state looks unsafe, with clear error responses for operators
- Database migrations now add the new rebuild tracking columns for existing installs
- Test coverage was expanded for lock state checks, heartbeat updates, failure detection, recovery decisions, and migration behavior

### Impact
`✅ Fewer unsafe rebuild starts`
`✅ Clearer rebuild recovery decisions`
`✅ More accurate rebuild status after restart`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/6c3ca414-ebd8-43a4-bc22-148ce414b848?refs=a1546d701461a5c778675dcdcbc710b2bc1a2dc5&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: src/logic/rebuild_failure_detection.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    if detect_orphaned_running_state(job, runtime_has_active_executor):
+        return RebuildInconsistency(
+            inconsistency_type=InconsistencyType.ORPHANED_RUNNING,
+            job_id=job.job_id,
+            reason=(f"Job {job.job_id} is 'running' in DB " f"but no active executor in runtime"),
+            detected_at=current_time,
+        )
+
+    # Check for crash suspicion (second priority)
+    if detect_crash_suspicion(job, heartbeat_threshold, now=current_time):
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/rebuild_failure_detection.py]: Consolidated both tests to run under the ORPHANED_RUNNING umbrella into a clear unified check that captures both ways

File: src/logic/rebuild_failure_detection.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    Args:
+        job: The rebuild job to check.
+        ttl_seconds: Time-to-live threshold in seconds.
+        now: Current timestamp (for testing); defaults to utcnow().
+
+    Returns:
+        True if ownership is stale, False otherwise.
+    """
+    if job.status != "running":
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/rebuild_failure_detection.py]: I have replaced RebuildJobStatus enum in all related checks in the file

File: api/metrics.py [hunk 3]
```diff
@@ -37,8 +37,57 @@
...
+    """
+    status_map = {
+        "none": 0,
+        "pending": 1,
+        "running": 2,
+        "succeeded": 3,
+        "failed": 4,
+        "cancelled": 5,
+    }
+    REBUILD_STATE_STATUS.set(status_map.get(status, 0))
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/metrics.py]: Updated the mapping of rebuild job statuses to numeric values. Introduced a dedicated 'unknown' value for better alerting.

File: src/logic/rebuild_failure_detection.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        return False
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/rebuild_failure_detection.py]: While RebuildJobStatus(str, Enum) being a string-based enum made both patterns work functionally, the inconsistency created:

   - Type safety issues
   - Cognitive load for developers
   - Tests that masked the pattern divergence

  The codebase now consistently uses enum instances for comparisons (matching repository.py), reserving .value only for dictionary keys, logging, and external boundaries.

File: tests/unit/test_rebuild_failure_detection.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+
+@pytest.fixture
+def running_job():
+    """Create a running rebuild job for testing."""
+    now = datetime.now(timezone.utc)
+    return RebuildJobORM(
+        job_id="test-job-1",
+        requested_by="operator@example.com",
+        status="running",
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_rebuild_failure_detection.py]: Fixed replacing strings with enums

File: src/logic/rebuild_failure_detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/rebuild_failure_detection.py]: Fixed

File: src/logic/rebuild_failure_detection.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+    if job.last_heartbeat_at is None:
+        # Running without any heartbeat is stale
+        return True
+
+    current_time = now or datetime.now(timezone.utc)
+    age = current_time - job.last_heartbeat_at
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/rebuild_failure_detection.py]: Fixed

File: api/metrics.py [hunk 7]
```diff
@@ -37,8 +37,68 @@
...
+    # Use a dedicated 'unknown' value (e.g., -1) that triggers alerts
+    # rather than defaulting to 0 (none).
+    gauge_value = mapping.get(status.lower())
+
+    if gauge_value is None:
+        logger.error(f"Inconsistency Detected: Received unknown job status '{status}'. Mapping to UNKNOWN_STATE (-1).")
+        return -1
+
+    return gauge_value
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/metrics.py]: Fixed

File: api/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mohavro) [reply api/metrics.py]: Fixed

File: .github/workflows/ci-gate-spec.yaml [hunk 8]
```diff
@@ -0,0 +1,165 @@
+name: Stage 5C - Distributed Coordination Safety Gates
+
+on:
+  pull_request:
+  push:
+    branches: [ main ]
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .github/workflows/ci-gate-spec.yaml]: Fixed by moving the workflow file to the .github/workflows directory

File: tools/ci/check_coordination_invariants.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    # Duplicate execution control paths (heuristic)
+    "DUPLICATE_EXECUTION_PATH": re.compile(
+        r"(execute_rebuild).*?(execute_rebuild)",
+        re.DOTALL,
+    ),
+    # Missing explicit gating usage (heuristic)
+    "MISSING_RECOVERY_GATE": re.compile(
+        r"execute_rebuild(?!.*RecoveryGate)",
+        re.DOTALL,
+    ),
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tools/ci/check_coordination_invariants.py]: @coderabbit can you suggest code to fix this?

**Review by mohavro — COMMENTED** [review]
Approved with follow-up findings.

Rationale:

the coordination model improved materially,
the scope creep stayed inside the correct architectural domain,
runtime integrity is stronger after this PR than before it,
and reverting/decomposing the added gating logic would likely reduce coherence more than improve sequencing purity.

Most importantly:
the PR continues the project’s strongest architectural trend:

explicit operational semantics replacing implicit runtime behaviour.

That is the correct direction for a coordination-aware system.

File: api/routers/graph_admin.py [hunk 10]
```diff
@@ -743,6 +754,15 @@ def _perform_rebuild_and_persist_sync(
...
 
         lock_acquired = True
+
+        gate = RecoveryGate(
+            session_factory=session_factory,
+            lock=dist_lock,
+            runtime_has_active_executor=False,
+            lock_ttl_seconds=lock_ttl,
+        )
+        gate.ensure_safe_to_execute()
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/routers/graph_admin.py:868]: The PR description should be updated.

**Comment by mohavro** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/1167#pullrequestreview-4298217578)

File: src/data/repository.py [hunk 11]
```diff
@@ -1201,3 +1204,120 @@ def release_distributed_lock(self, *, lock_name: str, holder_id: str) -> None:
...
+            ValueError: If the job does not exist or is not in running status.
+        """
+        job = self.session.get(RebuildJobORM, job_id)
+        if job is None:
+            raise ValueError(f"Rebuild job {job_id} not found")
+        if job.status != RebuildJobStatus.RUNNING:
+            current_status = job.status.value if isinstance(job.status, RebuildJobStatus) else job.status
+            raise ValueError(
+                f"Cannot update heartbeat for job {job_id} " f"with status {current_status} (must be running)"
+            )
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/data/repository.py:1325]:
```suggestion
            raise ValueError(
                f"Cannot update heartbeat for job {job_id} with status {current_status} (must be running)"
            )
```

**Comment by mohavro** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/1167#pullrequestreview-4300703676)

File: src/logic/recovery_gate.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            decision = RecoveryDecision(
+                action=decision.action,
+                reason=(
+                    "Orphaned running rebuild detected with owner mismatch: "
+                    f"job worker_id={job.active_worker_id!r}, "
+                    f"current lock holder_id={self.lock.holder_id!r}"
+                ),
+                inconsistency_type=decision.inconsistency_type,
+                safe_to_execute=decision.safe_to_execute,
+            )
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/recovery_gate.py]:
```suggestion
        if owner_mismatch:
            from src.logic.rebuild_recovery import RecoveryDecision

            decision = RecoveryDecision(
                action=RecoveryAction.RESET,
                reason=(
                    "Orphaned running rebuild detected with owner mismatch: "
                    f"job worker_id={job.active_worker_id!r}, "
                    f"current lock holder_id={self.lock.holder_id!r}"
                ),
                inconsistency_type=decision.inconsistency_type,
                safe_to_execute=decision.safe_to_execute,
            )
```

File: api/metrics.py [hunk 13]
```diff
@@ -37,8 +43,136 @@
...
+        )
+        update_rebuild_state_metric("unknown")
+    except Exception as exc:  # noqa: BLE001
+        # DB read failure - fail gracefully
+        logger.exception(
+            "Failed to initialize rebuild state metric from DB: %s (%s). Setting to unknown.",
+            type(exc).__name__,
+            str(exc),
+        )
+        update_rebuild_state_metric("unknown")
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/metrics.py:195]: Good catch. Using logger.exception has indeed caused fairly significant stack trace problems. 

The same driver to revert from logger.error back to logger.exception persists in that sonar is flagging this on CI.

Given stack trace issues were in the past and accepting that change could resurface problems, considering the changes in the code recently is the risk actually still present (from using logger.exception) and or has it been mitigated fully or partially through other means?

**Comment by mohavro** [comment]
@codeant-ai review

**Comment by mohavro** [comment]
@codex review pr

**Comment by mohavro** [comment]
@coderabbitai review

**Comment by mohavro** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/1167#pullrequestreview-4303870650)

**Review by mohavro — COMMENTED** [review]
# PR #1167 Review: Stage 5C.1 - Failure Detection and Recovery State Model

## Executive Summary

**Status**: ✅ **APPROVED** - PR objectives met with strong foundational implementation

**Verdict**: The PR successfully implements Stage 5C.1 objectives and introduces necessary early 5C.2 primitives. The foundational logic is sound, well-tested, and establishes coordination-first execution semantics.

---

## Architectural Milestone Achieved

**This PR marks a critical transition to coordination-first execution semantics:**

- **Execution is no longer authoritative** - runtime state must conform to coordination state
- **Database state and coordination logic are authoritative** - execution requests are validated against DB truth
- **Recovery gate enforces coordination boundaries** - unsafe execution is blocked, not just warned

This architectural shift is a prerequisite for:
- Safe crash recovery (5C.2)
- Deterministic cancellation (5C.3)
- Multi-worker correctness under failure

**Without this transition, Stage 5C.2 and 5C.3 would not be safely implementable.**

---

## Compliance with Issue #1166 PR 5C.1 Checklist

### ✅ Required Components (All Implemented)

1. **Persistent Rebuild State Model** ✅
   - [`active_worker_id`](src/data/db_models.py:250) and [`last_heartbeat_at`](src/data/db_models.py:251-254) columns added
   - Idempotent migrations for SQLite ([`migrations.py`](src/data/migrations.py:124-153)) and PostgreSQL ([`migrations.py`](src/data/migrations.py:155-178))
   - Status field already existed; heartbeat tracking added as specified

2. **Failure Detection Layer** ✅
   - [`detect_stale_ownership()`](src/logic/rebuild_failure_detection.py:49-82): Detects heartbeat TTL expiry
   - [`detect_orphaned_running_state()`](src/logic/rebuild_failure_detection.py:85-106): Detects DB/runtime divergence
   - [`detect_crash_suspicion()`](src/logic/rebuild_failure_detection.py:109-146): Detects missing/stale heartbeats
   - [`detect_rebuild_inconsistency()`](src/logic/rebuild_failure_detection.py:241-297): Main service function

3. **Recovery Decision Function** ✅
   - [`determine_recovery_action()`](src/logic/rebuild_recovery.py:39-170): Returns `RESUME | RESET | WAIT | UNSAFE`
   - Deterministic outputs verified by [tests](tests/unit/test_rebuild_recovery.py:75-86)
   - Clear decision logic with proper prioritization

4. **Observability Metrics** ✅
   - [`REBUILD_STATE_STATUS`](api/metrics.py:47-50): Gauge metric for rebuild state
   - [`REBUILD_RECOVERY_TRIGGERS`](api/metrics.py:52-56): Counter for inconsistency detection
   - [`initialize_rebuild_state_metric_from_db()`](api/metrics.py:123-198): Startup initialization

5. **Minimal Recovery-Side Mutation Introduced (RESET only)** ✅
   - [`RecoveryGate.ensure_safe_to_execute()`](src/logic/recovery_gate.py:260-304) includes automatic RESET recovery
   - **Not pure detection** - performs bounded, idempotent state mutation
   - **Structurally required**: Pure detection model is non-viable once execution gating is introduced
   - Without minimal reconciliation capability, system becomes permanently blocked
   - RESET logic is bounded: marks orphaned job as failed, requires lock reacquisition

---

## Scope Analysis

### ✅ Core 5C.1 Scope (Correctly Implemented)

- Database schema additions
- Failure detection logic
- Recovery decision logic
- Observability metrics
- Migration infrastructure

### ✅ Early 5C.2 Primitives (Intentionally Included)

**RecoveryGate Execution Logic** ([`recovery_gate.py:260-357`](src/logic/recovery_gate.py:260-357))

**Analysis**:
- **What**: Automatic RESET recovery for orphaned jobs
- **Why necessary**: A pure detection model becomes permanently blocked once execution gating is introduced. Minimal reconciliation capability is structurally required.
- **Risk**: Low - operation is idempotent, properly guarded by lock reacquisition, and bounded to RESET only
- **Implication for 5C.2**: Part of recovery execution is already solved; 5C.2 should not reimplement this

**Verdict**: This is not scope creep - it is the minimum viable recovery execution required for the gating model to function.

---

## Cross-Worker Safety Guarantees

**Achieved protections** ([`recovery_gate.py:96-186`](src/logic/recovery_gate.py:96-186)):

1. **Prevents concurrent execution**: Multiple workers cannot execute simultaneously
2. **Prevents stale-owner mutation**: Expired local lock cannot reset healthy remote worker
3. **Prevents unsafe remote reset**: Fresh heartbeat from different worker → UNSAFE (blocks reset)

**Implementation**:
- Owner mismatch detection with heartbeat staleness check
- Fresh heartbeat (< TTL) from different worker → UNSAFE decision
- Stale/missing heartbeat from different worker → RESET decision (safe recovery)

This is stronger than typical "split-brain prevention" - it provides deterministic safety under heartbeat ambiguity.

---

## Technical Review Findings

### ✅ Excellent Implementations

1. **Deterministic State Machine**
   - Recovery decisions are pure functions with no side effects
   - Same inputs always produce same outputs ([test coverage](tests/unit/test_rebuild_recovery.py:78-86))
   - Clear state transition rules

2. **Robust Heartbeat Mechanism**
   - Atomic conditional updates prevent race conditions ([`repository.py:1289-1305`](src/data/repository.py:1289-1305))
   - Worker ownership enforced via database constraints
   - Timezone-aware timestamp handling ([`rebuild_failure_detection.py:38-46`](src/logic/rebuild_failure_detection.py:38-46))

3. **Lock State Checking** ([`distributed_lock.py`](src/data/distributed_lock.py:18-24))
   - Explicit `LockState` enum: `VALID | EXPIRED | UNKNOWN | LOST`
   - Clear semantics for each state

4. **Heartbeat Keeper Thread** ([`graph_admin.py:669-712`](api/routers/graph_admin.py:669-712))
   - Prevents lock expiry during long rebuilds
   - Signals main thread on lock loss
   - Proper cleanup in finally block

5. **Metric Initialization** ([`api/metrics.py:123-198`](api/metrics.py:123-198))
   - Restores state from DB on startup
   - Handles multiple running jobs gracefully
   - Bounded error logging

6. **Security & Safety**
   - Migration whitelist prevents SQL injection ([`migrations.py:26-31`](src/data/migrations.py:26-31))
   - Bounded error messages prevent credential leakage
   - Fail-closed on heartbeat write failure ([`graph_admin.py:637-664`](api/routers/graph_admin.py:637-664))

### ⚠️ Minor Observations

1. **CI Safety Gates** ([`check_coordination_invariants.py`](tools/ci/check_coordination_invariants.py))
   - Static invariant checker enforces coordination patterns
   - Uses AST-based analysis for ownership mutation detection ([lines 111-128](tools/ci/check_coordination_invariants.py:111-128))
   - Uses AST-based analysis for `ensure_safe_to_execute()` call detection ([lines 94-108](tools/ci/check_coordination_invariants.py:94-108))
   - **Strength**: Prevents bypass via comments/strings, catches spacing variants
   - **Current state**: Pattern-based for execution entrypoints, AST-based for critical invariants

2. **Migration 002 Superseded**
   - [`002_add_heartbeat_columns.sql`](migrations/002_add_heartbeat_columns.sql) is a no-op
   - Columns added directly to 001_initial.sql in commit 250925f6
   - Python wrapper handles both cases correctly
   - Well-documented in SQL comments

3. **Timezone Handling**
   - Naive datetime normalization assumes UTC ([`rebuild_failure_detection.py:38-46`](src/logic/rebuild_failure_detection.py:38-46))
   - **Risk**: Low - SQLite stores as TEXT, PostgreSQL uses TIMESTAMPTZ
   - **Recommendation**: Acceptable for current scope; assumption is documented

4. **Worker ID Length Validation**
   - Enforced at application layer ([`repository.py:1268-1269`](src/data/repository.py:1268-1269))
   - **Recommendation**: Consider DB constraint in future migration

---

## Test Coverage Assessment

### ✅ Strong Coverage

- **Failure Detection**: 100% of detection functions tested with boundary conditions
- **Recovery Logic**: Determinism verified, all state combinations tested
- **RecoveryGate**: Lock states, inconsistency handling, metric increments tested
- **Repository**: Heartbeat updates, ownership conflicts, multiple running jobs tested
- **Migrations**: Idempotency verified for both SQLite and PostgreSQL

### Deferred to 5C.2 (Per Issue #1166)

- Integration tests for multi-worker scenarios
- Failure injection tests
- Concurrent restart storm simulation

---

## Alignment with Repository Policies

### ✅ Compliant

- **Production Architecture**: Changes target FastAPI backend (production path)
- **PR Scope Guardrails**: Single primary decision (coordination-first architecture)
- **High-Risk Change Control**: Database changes properly scoped with migrations
- **Test Coverage**: Comprehensive unit tests for all new logic
- **Security**: Bounded logging, migration whitelist, fail-closed on errors

---

## Recommendations

### For Merge

1. ✅ **Approve and merge** - all objectives met, architectural milestone achieved
2. Document that RecoveryGate includes minimal RESET execution (structurally required, not scope creep)
3. Ensure 5C.2 PR does not reimplement RESET recovery (already solved here)

### For 5C.2 (Recovery Gate + Lock Failure Hardening)

**Critical: Do not reimplement what 5C.1 already provides:**

1. **RESET recovery is complete** - 5C.2 should focus on WAIT/UNSAFE handling
2. **Lock state classification exists** - extend, don't replace
3. **Heartbeat mechanism is operational** - 5C.2 adds failure handling, not mechanism itself

**New work for 5C.2:**
1. Lock refresh failure handling (currently logs but continues)
2. Explicit WAIT state resolution logic
3. UNSAFE state operator guidance
4. Multiple running jobs recovery path

### For 5C.3 (Executor Crash Recovery + Cancellation)

1. Idempotent execution contract
2. Cancellation propagation through heartbeat keeper
3. Terminal state enforcement
4. Stale executor mutation prevention

### For Documentation

1. Add ADR documenting coordination-first architecture transition
2. Document heartbeat mechanism and TTL semantics
3. Create runbook for operators on interpreting recovery metrics
4. Document timezone assumptions for naive datetime handling

---

## Final Verdict

**✅ APPROVED FOR MERGE**

PR #1167 successfully implements Stage 5C.1 objectives and establishes the coordination-first execution model required for safe distributed operation.

**Key Achievements**:
- ✅ Authoritative database state model with heartbeat tracking
- ✅ Deterministic failure detection and recovery decision logic
- ✅ Cross-worker safety: prevents concurrent execution, stale-owner mutation, and unsafe remote reset
- ✅ Minimal viable recovery execution (RESET only, structurally required)
- ✅ Comprehensive observability metrics with startup reconciliation
- ✅ Strong test coverage with boundary condition verification
- ✅ Proper security and safety guardrails

**Architectural Milestone**: Transition from execution-driven to coordination-driven system complete. This is the foundation that makes 5C.2 and 5C.3 safely implementable.

**Critical Note for 5C.2**: Do not reimplement RESET recovery or heartbeat mechanism - these are complete. Focus on lock failure hardening, WAIT/UNSAFE resolution, and operator guidance.
